proposed/activity-improvements.md

```diff
@@ -0,0 +1,393 @@
+    public enum ActivityCreationOptions
+    {
+        /// Don't allow creating the Activity object.
+        NoCreate,
+
+        /// Create the Activity but drop recording events and tags.
+        NoDataAllowed,
+
+        /// Create the Activity and allow recording events and tags
+        AllowData,
+
+        /// Create the Activity, allow recording events and tags, and turn on ActivityTraceFlags.Recorded (i.e. Allow Sampling).
+        AllowDataAndSampling
+    }
+}
+```
+
+### ActivitySource
+
+```C#
+
+namespace System.Diagnostics
+{
+    /// ActivitySource is the type that help create and start Activity and allow listening to the Activity events too.
+    public sealed class ActivitySource : IDisposable
+    {
+        /// Construct an ActivitySource object with the input name and component version
+        public ActivitySource(string name, Version version);
+
+        /// Returns the ActivitySource name.
+        public string Name { get; }
+
+        /// Returns the ActivitySource version.
+        public Version Version { get; }
+
+        /// Creates a new Activity object if there is any listener to the Activity, returns null otherwise.
+        public Activity? StartActivity(string name, ActivityKind kind = ActivityKind.Internal);
+
+        public Activity? StartActivity(string name, ActivityKind kind, ActivityContext context, IEnumerable<KeyValuePair<string, string?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default);
+
+        /// Dispose the ActivitySource object and remove the current instance from the global list. empty the listeners list too.
+        public void Dispose();
+
+        /// Add a listener to the Activity starting and stopping events.
+        public IDisposable AddActivityListener(
+                                Func<ActivitySource, string, ActivityKind, ActivityContext, IEnumerable<KeyValuePair<string, string?>>?, IEnumerable<ActivityLink>?, ActivityCreationOptions> getActivityCreationOption,
+                                Action<Activity> onActivityStarted,
+                                Action<Activity> onActivityStopped);
+
+        /// Add a listener to listen to ActivitySource creation events.
+        public static IDisposable AddListener(Action<ActivitySource> OnSourceAttaching);
+    }
+```
+
+### Activity
+
+```C#
+namespace System.Diagnostics
+{
+    public partial class Activity : IDisposable // added IDisposable
+    {
+        ....
+        public ActivityKind Kind { get; set; } = ActivityKind.Internal;
+        public string DisplayName { get; set; }
+        public ActivitySource Source { get; }
+
+        public Activity AddEvent(ActivityEvent e);
+        public IEnumerable<ActivityEvent> Events { get; }
+
+        public Activity AddLink(ActivityLink link);
+        public IEnumerable<ActivityLink> Links { get; }
+
+        public bool TryGetContext(out ActivityContext context);
```

Comment:
Maybe I am misremembering but I thought we were going to do this as a property that returns default(ActivityContext) if the format isn't W3C? A few reasons to do it that way:

a) This is the pattern TraceId, SpanId, ParentSpanId and ActivityFlags already follow.
b) The prediction over time is that nobody will use request id format so if we have two coding patterns that could work I think its fine to optimize for the case where W3C is the format being used.
c) if the dev isn't sure what the format is it isn't much different writing:
````
if(activity.TryGetContext(out ActivityContext ctx))
{
    foo(ctx);
}
````
vs.
````
if(activity.IdFormat == Format.W3C)
{
    foo(activity.ActivityContext);
}
````

Reply:
I'll fix that as you suggested.

Reply:
done